tchannel/tornado/peer.py

```diff
@@ -698,3 +482,167 @@ def clean_up_outgoing_request(request, connection, error):
+    def lookup(self, hostport):
+        """Look up a Peer for the given host and port.
+
+        Returns None if a Peer for the given host-port does not exist.
+        """
+        assert hostport, "hostport is required"
+        return self._peers.get(hostport, None)
+
+    def remove(self, hostport):
+        """Delete the Peer for the given host port.
+
+        Does nothing if a matching Peer does not exist.
+
+        :returns: The removed Peer
+        """
+        assert hostport, "hostport is required"
+        return self._peers.pop(hostport, None)
+
+    def add(self, peer):
+        """Add an existing Peer to this group.
+
+        A peer for the given host-port must not already exist in the group.
+        """
+        assert peer, "peer is required"
+
+        if isinstance(peer, basestring):
+            # Assume strings are host-ports
+            peer = self.peer_class(self.tchannel, peer)
+
+        assert peer.hostport not in self._peers, (
+            "%s already has a peer" % peer.hostport
+        )
+
+        self._peers[peer.hostport] = peer
+
+    @property
+    def hosts(self):
+        """Get all host-ports managed by this PeerGroup."""
+        return self._peers.keys()
+
+    @property
+    def peers(self):
+        """Get all Peers managed by this PeerGroup."""
+        return self._peers.values()
+
+    def request(self, service, hostport=None, **kwargs):
+        """Initiate a new request through this PeerGroup.
+
+        :param hostport:
+            If specified, requests will be sent to the specific host.
+            Otherwise, a known peer will be picked at random.
+        :param service:
+            Name of the service being called. Defaults to an empty string.
+        """
+        return PeerClientOperation(
+            peer_group=self,
+            service=service,
+            hostport=hostport,
+            **kwargs)
+
+    def _connected_peers(self, hostports):
+        for hostport in hostports:
+            peer = self.get(hostport)
+            if peer.connected:
+                yield peer
+
+    def choose(self, hostport=None, blacklist=None):
+        """Choose a Peer that matches the given criteria.
+
+        :param hostport:
+            Specifies that the returned Peer must be for the given host-port.
+            Without this, all peers managed by this PeerGroup are
+            candidates.
+        :param blacklist:
+            Peers on the blacklist won't be chosen.
+        :returns:
+            A Peer that matches all the requested criteria or None if no such
+            Peer was found.
+        """
+
+        blacklist = blacklist or set()
+        if hostport:
+            return self.get(hostport)
+
+        hosts = self._peers.viewkeys() - blacklist
+        if not hosts:
+            return None
+
+        peers = list(self._connected_peers(hosts))
+        if peers:
+            return peers[random.randint(0, len(peers)-1)]
+        else:
+            hosts = list(hosts)
```

Comment:
I swapped both, `peers[..]` and `hosts[..]` with `random.sample`, dropping the `list(hosts)` for `hosts` (since `random.sample` doesn't care if hosts is a list or not). In multiple runs, the median time was approximately the same (with a variance of about 5-8 microseconds). The rest of the computation is too expensive for this to make any difference.